workers/price-reporter/src/manager/external_executor.rs

```diff
@@ -0,0 +1,467 @@
+                }
+
+                // Dequeue the next job from elsewhere in the local node
+                Some(job) = job_receiver.recv() => {
+                    if self.config.disabled {
+                        warn!("ExternalPriceReporter received job while disabled, ignoring...");
+                        continue;
+                    }
+
+                    tokio::spawn({
+                        let mut self_clone = self.clone();
+                        let msg_out_tx = msg_out_tx.clone();
+                        async move {
+                            if let Err(e) = self_clone.handle_job(job, msg_out_tx).await {
+                                error!("Error in ExternalPriceReporter execution loop: {e}");
+                            }
+                        }.instrument(info_span!("handle_job"))
+                    });
+                },
+
+                // Await cancellation by the coordinator
+                _ = cancel_channel.changed() => {
+                    info!("ExternalPriceReporter cancelled, shutting down...");
+                    return Err(PriceReporterError::Cancelled("received cancel signal".to_string()));
+                }
+            }
+        }
+    }
+
+    /// Spawns the task responsible for handling the websocket connection
+    /// with the external price reporter
+    fn spawn_ws_handler_loop(
+        &self,
+    ) -> (UnboundedSender<WebsocketMessage>, UnboundedReceiver<PriceMessage>) {
+        let price_reporter_url: Url = self
+            .config
+            .price_reporter_url
+            .clone()
+            .unwrap()
+            .parse()
+            .expect("Invalid price reporter URL");
+
+        let (msg_out_tx, msg_out_rx) = unbounded_channel();
+        let (msg_in_tx, msg_in_rx) = unbounded_channel();
+
+        let subscription_states = self.subscription_states.clone();
+
+        tokio::spawn(ws_handler_loop(
+            price_reporter_url,
+            subscription_states,
+            msg_out_rx,
+            msg_in_tx,
+        ));
+
+        (msg_out_tx, msg_in_rx)
+    }
+
+    /// Spawns the price streamer loop, which periodically publishes price
+    /// reports to the system bus for the given token pair
+    fn spawn_price_streamer_loop(self, base: Token, quote: Token) {
+        tokio::spawn(async move {
+            let topic_name = price_report_topic_name(&base, &quote);
+
+            loop {
+                if self.config.system_bus.has_listeners(&topic_name) {
+                    if let PriceReporterState::Nominal(report) =
+                        self.get_state(base.clone(), quote.clone()).await
+                    {
+                        self.config
+                            .system_bus
+                            .publish(topic_name.clone(), SystemBusMessage::PriceReport(report));
+                    }
+                }
+
+                tokio::time::sleep(Duration::from_millis(PRICE_REPORT_INTERVAL_MS)).await;
```

Comment:
nit: could be cleaner to construct the duration once at the function start, even if negligable cost it's a solid habit